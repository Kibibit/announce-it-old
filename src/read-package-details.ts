import fs from 'fs-extra';
import { get, isString } from 'lodash';
import { IBranchObject } from './announce-it';

export interface IPackageDetails {
  name: string;
  version: string;
  description: string;
  author: string;
  homepage: string;
  repository: {
    type: string;
    url: string;
  };
  announcements: {
    tweet: string;
    includeUnstable?: boolean;
  };
  release: {
    branches: Array<string | IBranchObject>;
  }
}

export function readPackageDetails(root: string): Promise<IPackageDetails> {
  return fs.readJson(`${ root }/package.json`)
    .then((packageDetails: Partial<IPackageDetails>) => {
      if (!isString(get(packageDetails, 'announcements.tweet'))) {
        throw new Error('no tweet template found. please see the readme for more details');
      }

      // here, we KNOW the package has the announcement template string
      return packageDetails as IPackageDetails;
    });
}
