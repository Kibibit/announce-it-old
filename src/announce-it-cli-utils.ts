#!/usr/bin/env node
import findRoot from 'find-root';
import { every, isString } from 'lodash';

import { KbAnnounceIt } from './announce-it';
import { readPackageDetails } from './read-package-details';

export class AnnounceItCli {
  areEnvVariablesDefined(env: NodeJS.ProcessEnv): Promise<void> {
    const envVariables = [
      'CONSUMER_KEY',
      'CONSUMER_SECRET',
      'ACCESS_TOKEN_KEY',
      'ACCESS_TOKEN_SECRET'
    ];

    const areEnvVariablesDefined = every(envVariables, (varName) => {
      return isString(env[ varName ]);
    });

    if (!areEnvVariablesDefined) {
      return Promise.reject(new Error(`These Environment variables are required: ${ envVariables.join(' ') }`));
    }

    return Promise.resolve();
  }

  findRoot(folder: string): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const root = findRoot(folder);

        resolve(root);
      } catch (err) {
        reject(new Error(`couldn't find npm project root: ${ err.message || err }`));
      }
    });
  }

  runAnnounceItCli(folder: string, env: NodeJS.ProcessEnv): Promise<string> {
    return this.findRoot(folder)
      .then((root) => readPackageDetails(root))
      .then((packageDetails) => {
        const announceIt = new KbAnnounceIt({
          consumerKey: env.CONSUMER_KEY as string,
          consumerSecret: env.CONSUMER_SECRET as string,
          accessTokenKey: env.ACCESS_TOKEN_KEY as string,
          accessTokenSecret: env.ACCESS_TOKEN_SECRET as string
        });

        return announceIt.announceRelease(packageDetails);
      });
  }
}
