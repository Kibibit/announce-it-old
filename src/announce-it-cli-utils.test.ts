import fs from 'fs-extra';
import { assign } from 'lodash';

import { AnnounceItCli } from './announce-it-cli-utils';
import { IPackageDetails } from './read-package-details';

jest.mock('find-root', () => (folder: string) => {
  if (folder === 'test-failure') {
    throw new Error('failure (test mock)');
  }

  return 'project-root';
});

const packageDetails: IPackageDetails = {
  name: 'package-name',
  description: 'package description',
  version: '1.0.0',
  author: 'test@test.com',
  homepage: 'pizza.com',
  repository: {
    type: '',
    url: ''
  },
  announcements: {
    tweet: 'nice!'
  }
};

describe('AnnounceItCli', () => {
  const announceItCli = new AnnounceItCli();

  describe('areEnvVariablesDefined', () => {
    it('should resolve if all required environment variables exists', () => {
      return expect(announceItCli.areEnvVariablesDefined({
        CONSUMER_KEY: 'CONSUMER_KEY',
        CONSUMER_SECRET: 'CONSUMER_SECRET',
        ACCESS_TOKEN_KEY: 'ACCESS_TOKEN_KEY',
        ACCESS_TOKEN_SECRET: 'ACCESS_TOKEN_SECRET'
      })).resolves.toMatchSnapshot();
    });
    it('should reject if at least 1 required environment variable is missing', () => {
      return expect(announceItCli.areEnvVariablesDefined({
        CONSUMER_KEY: 'CONSUMER_KEY',
        CONSUMER_SECRET: 'CONSUMER_SECRET',
        ACCESS_TOKEN_SECRET: 'ACCESS_TOKEN_SECRET'
      }))
        .rejects.toThrowErrorMatchingSnapshot()
        .then(() => expect(announceItCli.areEnvVariablesDefined({}))
          .rejects.toThrowErrorMatchingSnapshot());
    });
  });

  describe('findRoot', () => {
    it('should return the project root folder if ran inside npm project', () => {
      return expect(announceItCli.findRoot('test-folder')).resolves.toMatchSnapshot();
    });
    it('should throw error if not inside npm project', () => {
      return expect(announceItCli.findRoot('test-failure')).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('runAnnounceItCli', () => {
    it.todo('should run the entire flow on correct input');
    it.todo('should throw error on incorrect input');
  });
});
