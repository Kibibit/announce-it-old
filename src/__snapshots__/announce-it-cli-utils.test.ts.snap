// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnounceItCli areVariablesDefined should reject if at least 1 required variable is missing 1`] = `"These Variables are required: CONSUMER_KEY, CONSUMER_SECRET, ACCESS_TOKEN_KEY, ACCESS_TOKEN_SECRET, branch"`;

exports[`AnnounceItCli areVariablesDefined should reject if at least 1 required variable is missing 2`] = `"These Variables are required: CONSUMER_KEY, CONSUMER_SECRET, ACCESS_TOKEN_KEY, ACCESS_TOKEN_SECRET, branch"`;

exports[`AnnounceItCli areVariablesDefined should resolve if all required variables exists 1`] = `undefined`;

exports[`AnnounceItCli findRoot should return the project root folder if ran inside npm project 1`] = `"project-root"`;

exports[`AnnounceItCli findRoot should throw error if not inside npm project 1`] = `"couldn't find npm project root: failure (test mock)"`;
